### PR TITLE
Custom Errors Upgrade

### PR DESCRIPTION
**Pull request by JodsMigel**
Replaced all "require" to Custom Errors in:
- AprOracle.sol
- Factory.sol
- HardWorker.sol
- Platform.sol
- PriceReader.sol
- StrategyLogic.sol
- Swapper.sol
- VaultManager.sol
- VaultBase.sol
- RVaultBase.sol
- CVault.sol
- RMVault.sol
- RVault.sol
- Controllable.sol
- Zap.sol
- StrategyBase.sol
- GammaQuickSwapFarmStrategy.sol
- QuickswapV3StaticFarmStrategy.sol
- StrategyLib.sol
- StrategyProxy.sol
- Proxy.sol
- FactoryLib.sol
- AlgebraAdapter.sol
- ChainlinkAdapter.sol
- KyberAdapter.sol
- UniswapV3Adapter.sol

+ Fixed related tests

### CODE REVIEW

File: src/core/Factory.sol [hunk 1]
```diff
@@ -164,16 +164,24 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
     ) external nonReentrant returns (address vault, address strategy) {
         DeployVaultAndStrategyVars memory vars;
         vars.vaultConfig = vaultConfig[keccak256(abi.encodePacked(vaultType))];
-        require(vars.vaultConfig.implementation != address(0), "Factory: vault implementation is not available");
-        require(vars.vaultConfig.deployAllowed, "Factory: vault not allowed to deploy");
+        if(vars.vaultConfig.implementation == address(0)){
+            revert VaultImplementationIsNotAvailable();
```
Comment (a17) [inline src/core/Factory.sol:168]: add `string vaultType` arg

File: src/core/Factory.sol [hunk 2]
```diff
@@ -164,16 +164,24 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
+        }
+        if(!vars.vaultConfig.deployAllowed){
+            revert VaultNotAllowedToDeploy();
```
Comment (a17) [inline src/core/Factory.sol:171]: add `string vaultType` arg

File: src/core/Factory.sol [hunk 3]
```diff
@@ -164,16 +164,24 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
         vars.strategyIdHash = keccak256(bytes(strategyId));
         vars.platform = platform();
         vars.buildingPermitToken = IPlatform(vars.platform).buildingPermitToken();
         vars.buildingPayPerVaultToken = IPlatform(vars.platform).buildingPayPerVaultToken();
 
         StrategyLogicConfig storage config = strategyLogicConfig[vars.strategyIdHash];
-        require(config.implementation != address(0), "Factory: strategy implementation is not available");
-        require(config.deployAllowed, "Factory: strategy logic not allowed to deploy");
+        if(config.implementation == address(0)){
+            revert StrategyImplementationIsNotAvailable();
```
Comment (a17) [inline src/core/Factory.sol:180]: add `string strategyId` arg

File: src/core/Factory.sol [hunk 4]
```diff
@@ -164,16 +164,24 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
+        }
+        if(!config.deployAllowed){
+            revert StrategyLogicNotAllowedToDeploy();
```
Comment (a17) [inline src/core/Factory.sol:183]: add `string strategyId` arg

File: src/core/Factory.sol [hunk 5]
```diff
@@ -265,27 +282,39 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
+        if(vaultStatus[vault] != VaultStatusLib.ACTIVE){
+            revert NotActiveVault();
+        }
         IVaultProxy proxy = IVaultProxy(vault);
         bytes32 vaultTypeHash = proxy.VAULT_TYPE_HASH();
         address oldImplementation = proxy.implementation();
         address newImplementation = vaultConfig[vaultTypeHash].implementation;
-        require (vaultConfig[vaultTypeHash].upgradeAllowed, "Factory: upgrade denied");
-        require (oldImplementation != newImplementation, "Factory: already last version");
+        if(!vaultConfig[vaultTypeHash].upgradeAllowed){
```
Comment (a17) [inline src/core/Factory.sol:292]: add `bytes32 vaultTypeHash` arg

File: src/core/Factory.sol [hunk 6]
```diff
@@ -265,27 +282,39 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
+        }
         IStrategyProxy proxy = IStrategyProxy(strategyProxy);
         bytes32 idHash = proxy.STRATEGY_IMPLEMENTATION_LOGIC_ID_HASH();
         StrategyLogicConfig storage config = strategyLogicConfig[idHash];
         address oldImplementation = proxy.implementation();
         address newImplementation = config.implementation;
-        require (config.upgradeAllowed, "Factory: upgrade denied");
-        require (oldImplementation != newImplementation, "Factory: already last version");
+        if(!config.upgradeAllowed){
+            revert UpgradeDenied();
```
Comment (a17) [inline src/core/Factory.sol:313]: add `bytes32 strategyIdHash` arg

File: src/core/Factory.sol [hunk 7]
```diff
@@ -265,27 +282,39 @@ contract Factory is Controllable, ReentrancyGuardUpgradeable, IFactory {
...
+        }
+        if(oldImplementation == newImplementation){
+            revert AlreadyLastVersion();
```
Comment (a17) [inline src/core/Factory.sol:316]: add `bytes32 strategyIdHash` arg

**Review by a17 — CHANGES_REQUESTED** [review]

File: src/core/Factory.sol (hunk 1 above)
Comment (JodsMigel) [reply src/core/Factory.sol:168]: Not need to do

File: src/core/Factory.sol (hunk 2 above)
Comment (JodsMigel) [reply src/core/Factory.sol:171]: Not need to do

File: src/core/Factory.sol (hunk 3 above)
Comment (JodsMigel) [reply src/core/Factory.sol:180]: Not need to do

File: src/core/Factory.sol (hunk 4 above)
Comment (JodsMigel) [reply src/core/Factory.sol:183]: Not need to do

**Review by a17 — APPROVED** [review]

**Review by a17 — APPROVED** [review]

**Review by a17 — APPROVED** [review]

**Pull request merged**